### PR TITLE
MM 407 - Support user setting to avoid duplicate daily summaries

### PR DESCRIPTION
**Pull request by sibasankar-demansol**
Added `on_change`  for remainders in setting command to trigger daily update only if there is any change in the todo list of previous day

Ticket Link 
https://github.com/mattermost/mattermost-plugin-github/issues/407

### CODE REVIEW

**Comment by hanzei** [comment]
@sibasankar-demansol would you be open signing the [CLA](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pub?hl=en&hl=en&key=0&output=csv)?

**Comment by maisnamrajusingh** [comment]
@hanzei Siba here @sibasankar-demansol has already signed the agreement but it somehow doesn't seem to have transferred to the ci tool. Does he need to do anything else other than the agreement. I have forgotten how I got my PRs to be merged.

**Comment by sibasankar-umpteens** [comment]
> /check-cla

do i need to do this for every pull request , i have done this while creating other PR for this repo

**Comment by hanzei** [comment]
@sibasankar-umpteens You need to run `/check-cla` if you opened a PR _before_ signing the CLA. Hence you will only do that once.

File: server/plugin/command.go [hunk 1]
```diff
@@ -589,26 +598,31 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
 	mute.AddCommand(muteList)
 
 	settings := model.NewAutocompleteData("settings", "[setting] [value]", "Update your user settings")
-	setting := []model.AutocompleteListItem{{
-		HelpText: "Turn notifications on/off",
-		Item:     "notifications",
+	settingNotifications := model.NewAutocompleteData("notifications", "", "Turn notifications on/off")
+	settingValue := []model.AutocompleteListItem{{
+		HelpText: "Turn setting on",
```
Comment (hanzei) [inline server/plugin/command.go:603]: Given that there is now a per settings list of option: Should the help texts be more specific i.e.
```suggestion
		HelpText: "Turn notifications on",
```

File: server/plugin/command.go [hunk 2]
```diff
@@ -589,26 +598,31 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+		Item:     "on-change",
 	}}
-	settings.AddStaticListArgument("", true, value)
+	remainderNotifications.AddStaticListArgument("", true, settingValue)
+	settings.AddCommand(remainderNotifications)
 	github.AddCommand(settings)
-
 	issue := model.NewAutocompleteData("issue", "[command]", "Available commands: create")
-
 	issueCreate := model.NewAutocompleteData("create", "[title]", "Open a dialog to create a new issue in Github, using the title if provided")
```
Comment (hanzei) [inline server/plugin/command.go:626]: Unrelated change

File: server/plugin/command.go [hunk 3]
```diff
@@ -394,12 +394,21 @@ func (p *Plugin) handleSettings(_ *plugin.Context, _ *model.CommandArgs, paramet
...
+		}
 		value = true
-	} else if strValue != settingOff {
+		userInfo.Settings.OnChange = true
+	case strValue != settingOff:
+		if setting == "reminders" {
+			return "Invalid value. Accepted values are: \"on\" or \"off\" or \"on-change\" ."
+		}
 		return "Invalid value. Accepted values are: \"on\" or \"off\"."
 	}
```
Comment (hanzei) [inline server/plugin/command.go:413]: This is becoming very hard to read. Let's just `switch` based on `setting`.

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+
 func (p *Plugin) PostToDo(info *GitHubUserInfo) {
 	text, err := p.GetToDo(context.Background(), info.GitHubUsername, p.githubConnect(*info.Token))
 	if err != nil {
 		p.API.LogWarn("Failed to get todo text", "userID", info.UserID, "error", err.Error())
 		return
 	}
-
-	p.CreateBotDMPost(info.UserID, text, "custom_git_todo")
+	if info.Settings.DailyReminder {
```
Comment (hanzei) [inline server/plugin/plugin.go:418]: Should this check `OnChange`?

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -493,7 +541,6 @@ func (p *Plugin) GetToDo(ctx context.Context, username string, githubClient *git
 			text += getToDoDisplayText(baseURL, assign.GetTitle(), assign.GetHTMLURL(), "")
 		}
 	}
-
```
Comment (hanzei) [inline server/plugin/plugin.go:496]: Unrelated change

File: server/plugin/plugin.go [hunk 6]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+	p.API.LogWarn("checking previousSummary", "previousSummary", previousSummary)
+	if err != nil {
+		return false, err
+	}
+	if previousSummary == text {
+		return true, nil
+	}
+	return false, nil
+}
+func (p *Plugin) StoreDailySummaryText(userName, summaryText string) (bool, error) {
```
Comment (hanzei) [inline server/plugin/plugin.go:394]: Nit pick: spacing
```suggestion
}

func (p *Plugin) StoreDailySummaryText(userName, summaryText string) (bool, error) {
```

File: server/plugin/plugin.go [hunk 7]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+	if err := p.API.KVSet(userName+dailySummary, []byte(summaryText)); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+func (p *Plugin) GetDailySummaryText(userName string) (string, error) {
```
Comment (hanzei) [inline server/plugin/plugin.go:400]: Nit pick: spacing

```suggestion
}

func (p *Plugin) GetDailySummaryText(userName string) (string, error) {
```

File: server/plugin/plugin.go [hunk 8]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
 	}
 }
 
+func (p *Plugin) CheckISDuplicateSummary(userName, text string) (bool, error) {
+	p.API.LogWarn("checking CheckISDuplicateSummary", "userName", userName)
```
Comment (hanzei) [inline server/plugin/plugin.go:383]: Left over from debugging?

File: server/plugin/plugin.go [hunk 9]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+	return true, nil
+}
+func (p *Plugin) GetDailySummaryText(userName string) (string, error) {
+	summaryByte, err := p.API.KVGet(userName + dailySummary)
+	if err != nil {
+		return "", err
+	}
+	if summaryByte == nil {
+		return "", nil
+	}
```
Comment (hanzei) [inline server/plugin/plugin.go:407]: Why is this check needed?

File: server/plugin/plugin.go [hunk 10]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
 	}
 }
 
+func (p *Plugin) CheckISDuplicateSummary(userName, text string) (bool, error) {
```
Comment (hanzei) [inline server/plugin/plugin.go:382]: Why is `IS` capitalized ?

Comment (hanzei) [inline server/plugin/plugin.go:382]: Could `text` be a bit more self explaining?

File: server/plugin/plugin.go [hunk 11]
```diff
@@ -376,14 +379,59 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+		return "", err
+	}
+	if summaryByte == nil {
+		return "", nil
+	}
+	p.API.LogWarn("GetDailySummaryText", "summary text ", string(summaryByte))
+	return string(summaryByte), nil
+}
+
 func (p *Plugin) PostToDo(info *GitHubUserInfo) {
```
Comment (hanzei) [inline server/plugin/plugin.go:412]: 0/5 Maybe this should return an `error` and we let the caller deal with the logging.

File: server/plugin/plugin.go (hunk 6 above)
Comment (hanzei) [inline server/plugin/plugin.go:394]: Why does this return a `bool`?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for adding this handy feature  @sibasankar-demansol :+1:

File: server/plugin/plugin.go (hunk 6 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved

File: server/plugin/plugin.go (hunk 11 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go:412]: it returns err and text value which got stored

File: server/plugin/plugin.go (hunk 10 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved

File: server/plugin/plugin.go (hunk 9 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved , not needed

File: server/plugin/command.go (hunk 2 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go]: resolved

File: server/plugin/command.go (hunk 1 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go]: resolved , text are more specific

File: server/plugin/plugin.go [hunk 12]
```diff
@@ -236,6 +238,7 @@ type GitHubUserInfo struct {
 type UserSettings struct {
 	SidebarButtons string `json:"sidebar_buttons"`
 	DailyReminder  bool   `json:"daily_reminder"`
+	OnChange       bool   `json:"on_change"`
```
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved , text is more specific

File: server/plugin/command.go [hunk 13]
```diff
@@ -394,12 +394,21 @@ func (p *Plugin) handleSettings(_ *plugin.Context, _ *model.CommandArgs, paramet
...
 	strValue := parameters[1]
 	value := false
-	if strValue == settingOn {
+	switch {
+	case strValue == settingOn:
+		value = true
+	case strValue == settingOnChange:
+		if setting != "reminders" {
+			return "Invalid value. Accepted values are: \"on\" or \"off\"."
+		}
```
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/command.go]: resolved ,changed to  entire snippet of code to switch

File: server/plugin/api.go [hunk 14]
```diff
@@ -499,7 +498,6 @@ func (p *Plugin) getConnected(c *Context, w http.ResponseWriter, r *http.Request
 		EnterpriseBaseURL: config.EnterpriseBaseURL,
 		Organization:      config.GitHubOrg,
 	}
-
```
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/api.go]: resolved

File: server/plugin/command.go (hunk 3 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go]: resolved , changed to switch

File: server/plugin/plugin.go (hunk 4 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved , Sorry my bad it got changed before pushing 
it should be onChange

File: server/plugin/plugin.go (hunk 5 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved

File: server/plugin/plugin.go (hunk 6 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved

File: server/plugin/plugin.go (hunk 7 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved

File: server/plugin/plugin.go (hunk 8 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go]: resolved , yes removed that part

File: server/plugin/plugin.go [hunk 15]
```diff
@@ -376,13 +379,58 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
 	}
 
+	if info.Settings.DailyReminderOnChange {
+		isSameSummary, err := p.CheckIfDuplicateDailySummary(info.GitHubUsername, text)
+		if err != nil {
+			p.API.LogWarn("Failed to check the summary on change ", "userID", info.UserID, "error", err.Error())
+			return
+		}
+		if isSameSummary {
+			p.API.LogWarn(" Wont Trigger Daily Summary as the summary is Change ", "NO Summary ", true)
```
Comment (hanzei) [inline server/plugin/plugin.go:425]: That seems to spammy to me. Maybe `LogDebug` is fine, but even that seems to much to me.

**Review by hanzei — APPROVED** [review]
Great work :+1:

File: server/plugin/plugin.go (hunk 15 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go:425]: should i remove the message completely , no debug or warn message?

Comment (mickmister) [reply server/plugin/plugin.go:425]: Let's just remove the log message here

File: server/plugin/command.go [hunk 16]
```diff
@@ -391,20 +392,36 @@ func (p *Plugin) handleSettings(_ *plugin.Context, _ *model.CommandArgs, paramet
...
-	if setting != settingNotifications && setting != settingReminders {
+	settingValue := parameters[1]
+
+	switch setting {
+	case settingNotifications:
+		switch {
+		case settingValue == settingOn:
+			userInfo.Settings.Notifications = true
+		case settingValue == settingOff:
+			userInfo.Settings.Notifications = false
```
Comment (mickmister) [inline server/plugin/command.go:403]: This can be cleaned up by using the `settingValue` in the switch statement

```suggestion
		switch settingValue {
		case settingOn:
			userInfo.Settings.Notifications = true
		case settingOff:
			userInfo.Settings.Notifications = false
```

File: server/plugin/command.go [hunk 17]
```diff
@@ -589,22 +602,32 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+	remainderNotifications := model.NewAutocompleteData("reminders", "", "Turn notifications on/off")
+	settingValue = []model.AutocompleteListItem{{
+		HelpText: "Turn reminders on",
 		Item:     "on",
 	}, {
-		HelpText: "Turn setting off",
+		HelpText: "Turn reminders off",
 		Item:     "off",
+	}, {
+		HelpText: "Turn reminders On - But only get reminders if any change occur",
```
Comment (mickmister) [inline server/plugin/command.go:625]:
```suggestion
		HelpText: "Turn reminders on, but only get reminders if any changes have occurred since the previous day's reminder",
```

File: server/plugin/command.go [hunk 18]
```diff
@@ -589,22 +602,32 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
...
+		Item:     "on-change",
```
Comment (mickmister) [inline server/plugin/command.go:626]: Can we use the constants e.g. `settingOnChange` here?

File: server/plugin/plugin.go [hunk 19]
```diff
@@ -376,13 +379,58 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
+	}
+	if previousSummary == text {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (p *Plugin) StoreDailySummaryText(userName, summaryText string) error {
+	if err := p.API.KVSet(userName+dailySummary, []byte(summaryText)); err != nil {
```
Comment (mickmister) [inline server/plugin/plugin.go:395]: Does it make sense to store this using the Mattermost user ID instead of the github username?

File: server/plugin/plugin.go (hunk 11 above)
Comment (mickmister) [reply server/plugin/plugin.go:411]: I think hanzei is saying that `PostToDo` should return an `error`. Then the caller of `PostToDo` will log the error, to avoid several `LogWarn` calls in the same function.

**Review by mickmister — CHANGES_REQUESTED** [review]
Nice work @sibasankar-demansol :+1:

The core functionality looks good. I have some requests on code style and using the Mattermost user ID to store the reminder text, instead of using the GitHub username.

File: server/plugin/plugin.go (hunk 11 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go:411]: that part of code was an existing code , should i change as you are asking for

Comment (sibasankar-demansol) [reply server/plugin/plugin.go:411]: done

File: server/plugin/plugin.go (hunk 15 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go:425]: done

File: server/plugin/command.go (hunk 16 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:403]: done

File: server/plugin/command.go (hunk 17 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:625]: done

File: server/plugin/command.go (hunk 18 above)
Comment (sibasankar-demansol) [reply server/plugin/command.go:626]: sure

File: server/plugin/plugin.go (hunk 19 above)
Comment (sibasankar-demansol) [reply server/plugin/plugin.go:395]: yes we can

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks @sibasankar-demansol! I have one request to ensure `DailyReminderOnChange` gets set to `false` when other settings are being chosen.

cc @deepakdemiwal

File: server/plugin/command.go [hunk 20]
```diff
@@ -391,20 +392,36 @@ func (p *Plugin) handleSettings(_ *plugin.Context, _ *model.CommandArgs, paramet
...
+		}
+	case settingReminders:
+		switch settingValue {
+		case settingOn:
+			userInfo.Settings.DailyReminder = true
+		case settingOff:
+			userInfo.Settings.DailyReminder = false
+		case settingOnChange:
+			userInfo.Settings.DailyReminder = true
+			userInfo.Settings.DailyReminderOnChange = true
```
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/command.go:415]: As the type is Boolean it will be by default false , that's the reason i didn't added for other case

File: server/plugin/plugin.go [hunk 21]
```diff
@@ -376,14 +379,56 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) {
...
 
+	if info.Settings.DailyReminderOnChange {
+		isSameSummary, err := p.CheckIfDuplicateDailySummary(userID, text)
+		if err != nil {
+			return err
+		}
+		if isSameSummary {
+			return nil
+		}
+		err = p.StoreDailySummaryText(userID, text)
```
Comment (mickmister) [inline server/plugin/plugin.go:425]: @hanzei What do you think about storing the daily summary text inside the `GitHubUserInfo` struct, instead of using a separate KV row?

File: server/plugin/api.go [hunk 22]
```diff
@@ -529,7 +529,9 @@ func (p *Plugin) getConnected(c *Context, w http.ResponseWriter, r *http.Request
 		lt := time.Unix(lastPostAt/1000, 0).In(timezone)
 		if nt.Sub(lt).Hours() >= 1 && (nt.Day() != lt.Day() || nt.Month() != lt.Month() || nt.Year() != lt.Year()) {
 			if p.HasUnreads(info) {
-				p.PostToDo(info)
+				if err := p.PostToDo(info, c.UserID); err != nil {
+					p.API.LogWarn("Failed in PostToDo", "userID", info.UserID, "error", err.Error())
```
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/api.go:533]: Done

File: server/plugin/plugin.go (hunk 21 above)
Comment (mickmister) [reply server/plugin/plugin.go:425]: Actually that would make it so we are fetching summary from the KV store every time a user wants to do some action with the plugin. Probably fine to keep it as-is

File: server/plugin/command.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/command.go]: If I run `/github settings reminders on-change` then `/github settings reminders on`, that will not work without changing that to false right?

**Review by mickmister — APPROVED** [review]
LGTM, thanks @sibasankar-demansol!

File: server/plugin/command.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sibasankar-demansol) [reply server/plugin/command.go]: Yes seems this may have issue , anyway have made it false for other cases too
